Support host android apps with `kts` gradle files for add to app

gmackall opened the pull request:
Allows applying of `include_flutter.groovy` via the `apply from:` syntax, which allows using a host app that is using the Gradle Kotlin DSL (the default these days when creating an Android app in AS).

Explanation: The `include_flutter.groovy` script is currently not able to be called by Kotlin gradle files, because it is [intended to be invoked with the following lines](https://docs.flutter.dev/add-to-app/android/project-setup#depend-on-the-modules-source-code):
```
setBinding(new Binding([gradle: this]))                                // new
evaluate(new File(                                                     // new
    settingsDir.parentFile,                                            // new
    'flutter_module/.android/include_flutter.groovy'                   // new
))
```

`setBinding` isn't part of the Kotlin gradle DSL, and there isn't (that I can find) an easy Kotlin equivalent. If this binding isn't set, the reference to `gradle` in `include_flutter.groovy` is wrong, which breaks the script.

This PR modifies `include_flutter.groovy` to also support being invoked through the standard way of invoking a script via the Gradle Groovy/Kotlin DSLs, which is `apply from:` (or it's slightly different Kotlin syntax). The start of the script identifies which of the two approaches is being used by checking if the binding is set, and then initializes some variables differently depending on the case.

If we land this, I believe we should update the example Gradle files for both the `kts` and `groovy` cases to prefer the `apply from` syntax as I think this is the syntax most developers would be more familiar with already seeing in their Gradle files.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

mackall-work committed:
maybe try this

mackall-work committed:
deletion I forgot

gmackall commented:
cc @craiglabenz

bartekpacia reviewed (approved):
That's a very nice change!

LGTM

mackall-work committed:
new host app, move old host apps to unified dir

mackall-work committed:
modifications to host app to make it work

mackall-work committed:
rename tests, fix file paths. Still failing, figure out copy of read-only asset in test

mackall-work committed:
ci.yaml changes

mackall-work committed:
bringup

gmackall committed:
Merge branch 'master' into new_module_include_tmpl

gmackall commented:
converting to draft till I get the new test passing locally

mackall-work committed:
add test owner

mackall-work committed:
test working, remove webp files

mackall-work committed:
readmes

gmackall committed:
Merge branch 'master' into new_module_include_tmpl

gmackall commented:
Should be ready for review! Ended up making a number of changes for the testing:
1. Made a new directory `pure_android_host_apps` within `dev/integration_tests` to house all the android apps we use in integration testing. I think I moved all of the existing ones there, hopefully I didn't miss any
2. Added another host android app, created with the latest templates that Android Studio uses, and which uses the Kotlin DSL for its gradle files
3. Mostly copied the existing test that tested we could depend on a Flutter module as an AAR in a host Android app, to add a new test that tests we can do the same with that Flutter module as source (corresponding [to this option](https://docs.flutter.dev/add-to-app/android/project-setup#depend-on-the-modules-source-code))

gmackall commented:
I also isolated the changes that needed to be made to the new host app (that uses Kotlin gradle files) in a single commit, so here this is the diff that users will need to make:
https://github.com/flutter/flutter/pull/156502/commits/1f1e92ac79403745d609220ce74c09a446eb4e6c

mackall-work committed:
exclude host apps from linting, fix a couple of new lines cases

mackall-work committed:
another newline

mackall-work committed:
try to see if this runs on the bots successfully

bartekpacia commented on packages/flutter_tools/templates/module/android/library_new_embedding/include_flutter.groovy.copy.tmpl:
```diff
@@ -1,5 +1,13 @@
-def scriptFile = getClass().protectionDomain.codeSource.location.toURI()
-def flutterProjectRoot = new File(scriptFile).parentFile.parentFile
+def gradle = null
+def flutterProjectRoot = null
+if (!getBinding().getVariables().containsKey("gradle")) {
+    gradle = this
+    flutterProjectRoot = gradle.buildscript.getSourceFile().getParentFile().getParentFile().absolutePath
+} else {
+    gradle = getBinding().getVariables().get("gradle")
+    def scriptFile = getClass().protectionDomain.codeSource.location.toURI()
+    flutterProjectRoot = new File(scriptFile).parentFile.parentFile.absolutePath
+}
```
I think a comment here would be nice, describing when each case happens.

gmackall replied:
Good idea, added!

bartekpacia reviewed (approved):
I don't see anything wrong. LGTM

Really happy to see this change being made.

mackall-work committed:
comment

mackall-work committed:
remove commented code from AS app output template

gmackall commented on dev/integration_tests/pure_android_host_apps/host_app_kotlin_gradle_dsl/app/src/main/java/com/example/myapplication/ui/theme/Type.kt:
```diff
@@ -0,0 +1,34 @@
+        letterSpacing = 0.5.sp
+    )
+    */
+)
```
This was part of the output of the AS app templates, removed

gmackall commented:
Trying to figure out why the `ktlint` output is different on ci than locally - I can't seem to get it to properly exclude the `pure_android_host_apps` dir on CI, will land once I can get analysis passing

mackall-work committed:
rework editorconfig

mackall-work committed:
licenses

mackall-work committed:
newlines

bartekpacia commented on packages/flutter_tools/lib/src/android/gradle_utils.dart:
```diff
@@ -54,14 +54,14 @@ const String oneMajorVersionHigherJavaVersion = '23';
 // Supported here means supported by the tooling for
 // flutter analyze --suggestions and does not imply broader flutter support.
-const String maxKnownAndSupportedGradleVersion = '8.7';
+const String maxKnownAndSupportedGradleVersion = '8.10';
```
Are point releases significant here? I.e. will using `8.10.2` result in some warning?

gmackall replied:
Whoops! That was on the wrong branch, should be on https://github.com/flutter/flutter/pull/157032. Thanks for pointing it out!

gmackall replied:
I think the answer to the question by the way is that yes they are significant, and I'll want to figure out how we handle patch versions for that PR. 

I suspect that the correct way to handle them is to basically make the max known and supported `8.10.99` for example, because I don't think we will want to update that value for each patch version, and it's probably fine to just automatically assume new patch versions for a known minor version are supported.

But that is a question for that other PR!

mackall-work committed:
Revert "newlines"

This reverts commit 76baafd2568915dd3a0366bd230295689c5915e4.

mackall-work committed:
newlines, but actually just newlines this time
